spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplate.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * <p>
+ * The main eventhub component for sending to and consuming from event hub
+ *
+ * @author Warren Zhu
+ */
+public class EventHubTemplate implements EventHubOperation {
+
+    private static final Log LOGGER = LogFactory.getLog(EventHubTemplate.class);
+    private final ConcurrentHashMap<Tuple<String, String>, Set<Consumer<Iterable<EventData>>>>
+            consumersByNameAndConsumerGroup = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Tuple<String, String>, EventHubCheckpointer> checkpointersByNameAndConsumerGroup =
+            new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Tuple<String, String>, EventProcessorHost> processorHostsByNameAndConsumerGroup =
+            new ConcurrentHashMap<>();
+
+    private final EventHubClientFactory clientFactory;
+
+    public EventHubTemplate(EventHubClientFactory clientFactory) {
+        this.clientFactory = clientFactory;
+    }
+
+    @Override
+    public CompletableFuture<Void> sendAsync(String eventHubName, EventData message,
+            PartitionSupplier partitionSupplier) {
+        Assert.hasText(eventHubName, "eventHubName can't be null or empty");
+        Assert.notNull(message, "eventData can't be null");
+        try {
+            EventHubClient client = this.clientFactory.getOrCreateEventHubClient(eventHubName);
+
+            if (partitionSupplier == null) {
+                return client.send(message);
+            } else if (!Strings.isNullOrEmpty(partitionSupplier.getPartitionId())) {
+                return this.clientFactory.getOrCreatePartitionSender(eventHubName, partitionSupplier.getPartitionId())
+                                         .send(message);
+            } else if (!Strings.isNullOrEmpty(partitionSupplier.getPartitionKey())) {
+                return client.send(message, partitionSupplier.getPartitionKey());
+            } else {
+                return client.send(message);
+            }
+        } catch (EventHubRuntimeException e) {
+            LOGGER.error(String.format("Failed to send to '%s' ", eventHubName), e);
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+        }
+    }
+
+    @Override
+    public Checkpointer<EventData> getCheckpointer(String destination, String consumerGroup) {
+        return checkpointersByNameAndConsumerGroup.get(new Tuple<>(destination, consumerGroup));
+    }
+
+    @Override
+    public synchronized boolean subscribe(String destination, Consumer<Iterable<EventData>> consumer,
+            String consumerGroup) {
+        Tuple<String, String> nameAndConsumerGroup = new Tuple<>(destination, consumerGroup);
+        consumersByNameAndConsumerGroup.putIfAbsent(nameAndConsumerGroup, new CopyOnWriteArraySet<>());
+        boolean added = consumersByNameAndConsumerGroup.get(nameAndConsumerGroup).add(consumer);
+
+        if (!added) {
+            return false;
```

Comment:
Wonder for the user who calls this subscribe method, and returned false for an `added` failure operation, the user will have no idea why the operation failed, and will not be able to recover/process such situation, also the method itself leaves no hint why the subscribe failed.

Same as above, personally prefer exception over returning boolean status.

Reply:
Please check the comment, false is only returned when this consumer is already subscribed. So user doesn't need to recover, or some programming error existed.